exporters/otlp/src/otlp_http_client.cc

```diff
@@ -680,38 +776,143 @@ opentelemetry::sdk::common::ExportResult OtlpHttpClient::Export(
+        std::chrono::steady_clock::time_point::max() - std::chrono::steady_clock::now());
+    if (timeout >= max_timeout)
+    {
+      timeout = std::chrono::microseconds::zero();
+    }
+    max_timeout = std::chrono::duration_cast<std::chrono::microseconds>(
+        std::chrono::system_clock::time_point::max() - std::chrono::system_clock::now());
+    if (timeout >= max_timeout)
+    {
+      timeout = std::chrono::microseconds::zero();
+    }
+  }
 
+  // Wait for all the sessions to finish
+  std::unique_lock<std::mutex> lock(session_waker_lock_);
+  if (timeout > std::chrono::microseconds::zero())
+  {
+    session_waker_.wait(lock, [this] {
+      std::lock_guard<std::recursive_mutex> guard{session_manager_lock_};
+      return running_sessions_.empty();
+    });
+  }
+  else
+  {
+    session_waker_.wait_for(lock, timeout, [this] {
+      std::lock_guard<std::recursive_mutex> guard{session_manager_lock_};
+      return running_sessions_.empty();
+    });
+  }
+
+  while (cleanupGCSessions())
+    ;
   return true;
 }
 
+void OtlpHttpClient::ReleaseSession(
+    const opentelemetry::ext::http::client::Session &session) noexcept
+{
+  bool has_session = false;
+
+  {
+    std::lock_guard<std::recursive_mutex> guard{session_manager_lock_};
+
+    auto seesion_iter = running_sessions_.find(&session);
+    if (seesion_iter != running_sessions_.end())
+    {
+      // Move session and handle into gc list, and they will be destroyed later
+      gc_sessions_.emplace_back(std::move(seesion_iter->second));
+      running_sessions_.erase(seesion_iter);
+
+      has_session = true;
+    }
+  }
+
+  if (has_session)
+  {
+    session_waker_.notify_all();
+  }
+}
+
+void OtlpHttpClient::addSession(
+    std::shared_ptr<opentelemetry::ext::http::client::Session> session,
+    std::unique_ptr<opentelemetry::ext::http::client::EventHandler> &&event_handle) noexcept
+{
+  if (!session || !event_handle)
+  {
+    return;
+  }
+
+  opentelemetry::ext::http::client::Session *key = session.get();
+  ResponseHandler *handle = static_cast<ResponseHandler *>(event_handle.get());
+
+  handle->Bind(this, *key);
+
+  HttpSessionData &session_data = running_sessions_[key];
+  session_data.session.swap(session);
+  session_data.event_handle.swap(event_handle);
+
+  // Send request after the session is added
+  key->SendRequest(*handle);
+}
+
+bool OtlpHttpClient::cleanupGCSessions() noexcept
+{
+  std::lock_guard<std::recursive_mutex> guard{session_manager_lock_};
+  std::list<HttpSessionData> gc_sessions;
+  gc_sessions_.swap(gc_sessions);
+
+  for (auto &session_data : gc_sessions)
+  {
+    // FinishSession must be called with same thread and before the session is destroyed
+    if (session_data.session)
+    {
+      session_data.session->FinishSession();
+    }
+  }
+
+  return !gc_sessions_.empty();
+}
+
 bool OtlpHttpClient::isShutdown() const noexcept
 {
-  const std::lock_guard<opentelemetry::common::SpinLockMutex> locked(lock_);
   return is_shutdown_;
```

Comment:
Without lock this read would have race condition.

Reply:
`is_shutdown_` and `isShutdown()` are protected by `session_manager_lock_` now.